### PR TITLE
feat(spanner): set manual affinity incase of gRPC-GCP extenstion

### PR DESCRIPTION
**Pull request by harshachinta**
Earlier, gRPC-GCP maintains affinity using the [ApiConfig](https://github.com/googleapis/java-spanner/blob/main/google-cloud-spanner/src/main/resources/com/google/cloud/spanner/spi/v1/grpc-gcp-apiconfig.json) by looking into the gRPC messages to get an affinity key.

This PR sets the affinity manually to gRPC-GCP via call options. Reason - With multiplexed sessions we need to maintain transaction to channel affinity, and with regular sessions we need to maintain session to channel affinity.
Setting manual affinity will override the API config setting.

References:
1. Java client design and benchmarking - https://docs.google.com/document/d/1iQwAGCwvatdoGxvfucd0eFAysx6roGMxJsoyygX79NY/edit?resourcekey=0-feEwEwbAPxsA-QydfaG53A&tab=t.0
2. gRPC-GCP manual affinity design - https://docs.google.com/document/d/1NfnaCHpE6EzePU_IBTZHvevGizYTBtjuMfHlm9SZ70w/edit?tab=t.0

### CODE REVIEW

File: google-cloud-spanner/src/main/java/com/google/cloud/spanner/spi/v1/GapicSpannerRpc.java [hunk 1]
```diff
@@ -529,6 +531,10 @@ private static String parseGrpcGcpApiConfig() {
   private static GcpManagedChannelOptions grpcGcpOptionsWithMetrics(SpannerOptions options) {
     GcpManagedChannelOptions grpcGcpOptions =
         MoreObjects.firstNonNull(options.getGrpcGcpOptions(), new GcpManagedChannelOptions());
+    GcpChannelPoolOptions gcpChanelPoolOptions =
```
Comment (olavloite) [inline google-cloud-spanner/src/main/java/com/google/cloud/spanner/spi/v1/GapicSpannerRpc.java:534]:
```suggestion
    GcpChannelPoolOptions gcpChannelPoolOptions =
```

File: google-cloud-spanner/src/main/java/com/google/cloud/spanner/spi/v1/GapicSpannerRpc.java [hunk 2]
```diff
@@ -1950,7 +1957,17 @@ <ReqT, RespT> GrpcCallContext newCallContext(
...
       context = context.withChannelAffinity(Option.CHANNEL_HINT.getLong(options).intValue());
+
+      // Set channel affinity in gRPC-GCP. This is a no-op for GAX.
+      context =
+          context.withCallOptions(
+              context
+                  .getCallOptions()
+                  .withOption(
+                      GcpManagedChannel.AFFINITY_KEY,
+                      Option.CHANNEL_HINT.getLong(options).toString()));
```
Comment (olavloite) [inline google-cloud-spanner/src/main/java/com/google/cloud/spanner/spi/v1/GapicSpannerRpc.java:1970]: This channel hint can be any random Long, meaning that we might be using a very large range of different channel hints. That again means that we might be adding an unbounded set of channel hints to the map in grpc-gcp. We should add a computation here that makes sure that the set of hints is limited (e.g. do something like `hint % options.getNumChannels()`).

See https://github.com/googleapis/java-spanner/blob/e859b29ccf4e68b1ab62cffdd4cf197011ba9878/google-cloud-spanner/src/main/java/com/google/cloud/spanner/AbstractReadContext.java#L342

File: google-cloud-spanner/src/main/java/com/google/cloud/spanner/spi/v1/GapicSpannerRpc.java [hunk 3]
```diff
@@ -1950,7 +1953,17 @@ <ReqT, RespT> GrpcCallContext newCallContext(
       boolean routeToLeader) {
     GrpcCallContext context = GrpcCallContext.createDefault();
     if (options != null) {
+      // Set channel affinity in GAX. This is a no-op for gRPC-GCP.
       context = context.withChannelAffinity(Option.CHANNEL_HINT.getLong(options).intValue());
+
+      // Set channel affinity in gRPC-GCP. This is a no-op for GAX.
```
Comment (olavloite) [inline google-cloud-spanner/src/main/java/com/google/cloud/spanner/spi/v1/GapicSpannerRpc.java:1959]: nit: is it possible to only do this if grpc-gcp is being used? I know that it is a no-op for gax, and so does not affect anything in that way, but `context.withCallOptions(..)` always creates a new `CallContext` instance. That means that we reserve a small bit of additional memory for every RPC that we invoke that then needs to be cleaned up again afterwards. (And yes, that's a bit of a pre-emptive optimization, so if it is not possible or hard, then please ignore this comment, but if there's a simple if-condition that we can use, then let's do that.)

**Review by olavloite — APPROVED** [review]

File: google-cloud-spanner/src/main/java/com/google/cloud/spanner/spi/v1/GapicSpannerRpc.java (hunk 3 above)
Comment (harshachinta) [reply google-cloud-spanner/src/main/java/com/google/cloud/spanner/spi/v1/GapicSpannerRpc.java]: I have made changes to do it only when grpcgcp is enabled. Can you please take a look?

Comment (olavloite) [reply google-cloud-spanner/src/main/java/com/google/cloud/spanner/spi/v1/GapicSpannerRpc.java]: LGTM

**Pull request merged**